application/src/main/java/org/togetherjava/tjbot/logging/discord/DiscordLogForwarder.java

```diff
@@ -0,0 +1,228 @@
+ * <p>
+ * Although unlikely to hit, the class maximally buffers {@value MAX_PENDING_LOGS} logs until
+ * discarding further logs. Under normal circumstances, the class can easily handle high loads of
+ * logs.
+ */
+final class DiscordLogForwarder {
+    private static final Logger logger = LoggerFactory.getLogger(DiscordLogForwarder.class);
+
+    private static final int MAX_PENDING_LOGS = 10_000;
+    private static final int MAX_BATCH_SIZE = 10;
+    private static final int MAX_RETRIES_UNTIL_DISCARD = 3;
+    private static final int HTTP_STATUS_TOO_MANY_REQUESTS = 429;
+    private static final int HTTP_STATUS_OK_START = 200;
+    private static final int HTTP_STATUS_OK_END = 300;
+
+    private static final HttpClient CLIENT = HttpClient.newHttpClient();
+    private static final ObjectMapper JSON = new ObjectMapper();
+    private static final ScheduledExecutorService SERVICE =
+            Executors.newSingleThreadScheduledExecutor();
+
+    /**
+     * The Discord webhook to send logs to.
+     */
+    private final URI webhook;
+    /**
+     * Internal buffer of logs that still have to be forwarded to Discord. Actions are synchronized
+     * using {@link #pendingLogsLock} to ensure thread safety.
+     */
+    private final Queue<LogMessage> pendingLogs = new PriorityQueue<>();
+    private final Object pendingLogsLock = new Object();
+
+    /**
+     * If present, a rate limit has been hit and further requests must be made only after this
+     * moment.
+     */
+    @Nullable
+    private Instant rateLimitExpiresAt;
+    /**
+     * The amount of subsequent failed requests. Requests are tried
+     * {@value MAX_RETRIES_UNTIL_DISCARD} times until discarded. Resets to 0 once a request was
+     * successful.
+     */
+    private int currentRetries;
+
+    DiscordLogForwarder(URI webhook) {
+        this.webhook = webhook;
+
+        SERVICE.scheduleWithFixedDelay(this::processPendingLogs, 5, 5, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Forwards the given log message to Discord.
+     * <p>
+     * Logs are not immediately forwarded, but on a schedule. If the maximal buffer size of
+     * {@value MAX_PENDING_LOGS} is exceeded, logs are discarded.
+     *
+     * @param event the log to forward
+     */
+    void forwardLogEvent(LogEvent event) {
+        if (pendingLogs.size() >= MAX_PENDING_LOGS) {
+            logger.warn(LogMarkers.NO_DISCORD,
+                    """
+                            Exceeded the max amount of logs that can be buffered. \
+                            Logs are forwarded to Discord slower than they pile up. Discarding the latest log...""");
+            return;
+        }
+
+        LogMessage log = LogMessage.ofEvent(event);
+
+        synchronized (pendingLogsLock) {
+            pendingLogs.add(log);
+        }
+    }
+
+    private void processPendingLogs() {
+        try {
+            if (rateLimitExpiresAt != null) {
```

Comment:
could u just return if `rateLimitExpiresAt == null` ?

Reply:
thats a different thing

Reply:
The method has to do extra logic in all cases:
* if its not null, there is rate limitting and we either
  * have to return if its still active
  * reset it if its expired already
* otherwise we execute the actual forwarding

Reply:
i agree though, the nested if can be hard to comprehend at first sight. ill create a helper method

Reply:
pushed, have a check